Tests/RegexTests/CustomTests.swift

```diff
@@ -0,0 +1,113 @@
+import _StringProcessing
+import XCTest
+
+// A nibbler processes a single character from a string
+private protocol Nibbler:
+  MatchingCollectionConsumer, RegexProtocol
+where Consumed == String {
+  func nibble(_: Character) -> Match?
+}
+
+extension Nibbler {
+  // Default implementation, just feed the character in
+  func matchingConsuming(
+    _ consumed: Consumed,
+    in range: Range<Consumed.Index>
+  ) -> (upperBound: Consumed.Index, match: Match)? {
+    guard !range.isEmpty else {
+      // FIXME: Do we have a full story here?
```

Comment:
`range` should be allowed to be empty, and (when we add it) `currentIndex` should be allowed to be any index in `range.lowerBound...range.upperBound`. Otherwise you wouldn't be able to implement something like `$` using this interface.

Reply:
I think the interface we talked yesterday was something like 
```swift
func matchingConsuming(
    _ consumed: Consumed,
    startingAt index: Consumed.Index, /* currentIndex in @natecook1000's comment */
    in range: Range<Consumed.Index>
  ) -> (upperBound: Consumed.Index, match: Match)?
```

and so `_MatcherInterface` will have to be updated to take the `startingAt` index too, right?

Reply:
Yeah, `_MatcherInterface` will include an extra index parameter. I'll follow up with a PR that makes this change and introduces the `CustomRegexComponent` protocol